clippy_lints/src/methods/manual_str_repeat.rs

```diff
@@ -0,0 +1,95 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_context;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item, match_type};
+use clippy_utils::{is_expr_path_def_path, paths};
+use if_chain::if_chain;
+use rustc_ast::util::parser::PREC_POSTFIX;
+use rustc_ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self, Ty, TyS};
+use rustc_span::symbol::{sym, Symbol};
+
+use super::MANUAL_STR_REPEAT;
+
+enum RepeatKind {
+    Str,
+    String,
+    Char,
+}
+
+fn get_ty_param(ty: Ty<'_>) -> Option<Ty<'_>> {
+    if let ty::Adt(_, subs) = ty.kind() {
+        subs.types().next()
+    } else {
+        None
+    }
+}
+
+fn parse_repeat_arg(cx: &LateContext<'_>, e: &Expr<'_>) -> Option<RepeatKind> {
+    if let ExprKind::Lit(lit) = &e.kind {
+        match lit.node {
+            LitKind::Str(..) => Some(RepeatKind::Str),
+            LitKind::Char(_) => Some(RepeatKind::Char),
+            _ => None,
+        }
+    } else {
+        let ty = cx.typeck_results().expr_ty(e);
+        if is_type_diagnostic_item(cx, ty, sym::string_type)
+            || (is_type_lang_item(cx, ty, LangItem::OwnedBox) && get_ty_param(ty).map_or(false, TyS::is_str))
+            || (match_type(cx, ty, &paths::COW) && get_ty_param(ty).map_or(false, TyS::is_str))
+        {
+            Some(RepeatKind::String)
+        } else {
+            let ty = ty.peel_refs();
+            (ty.is_str() || is_type_diagnostic_item(cx, ty, sym::string_type)).then(|| RepeatKind::Str)
+        }
+    }
+}
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, method_name: Symbol, args: &[Expr<'_>]) {
+    let ctxt = expr.span.ctxt();
+    if_chain! {
+        if let [take_expr] = args;
+        if method_name.as_str() == "collect";
```

Comment:
```suggestion
        if method_name == sym!(collect);
```

Reply:
This is done both ways in clippy right now, near a 50/50 split. The string comparison is the more efficient version as `sym!` is just a wrapper around `Symbol::intern`.

Reply:
Why is it more efficient? This is just a question. I thought `Symbol::as_str` is slow because docs says so.

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/symbol/struct.Symbol.html#method.as_str
> This is a slowish operation because it requires locking the symbol interner.

Reply:
Both `as_str` and `intern` lock the interner. `as_str` is then just indexing into a `Vec`. `intern` involves a lookup in a hash table, which both hashes the string, and compares at least two strings for equality (assuming the string is found, which it most certainly will be in this case).

Reply:
I slightly prefer `as_str()` since it won't ever intern a string unnecessarily. I considered enforcing this in the internal `UNNECESSARY_SYMBOL_STR` lint, but decided it doesn't matter enough to lint.

Reply:
I got it , thanks!